### ISSUE TITLE
Standalone Editor CORS error in opencast 11.4

### ISSUE THREAD

**Issue by AL-THD** [opencast/opencast#3538]
**Describe the issue**
In opencast 11.4 the standalone Editor cannot load the waveform due to a CORS error:
XMLHttpRequest blocked by CORS policy as no header is present on the requested ressource.

Browser - console output:
```
Access to XMLHttpRequest at 'https://<presentation-node>/static/mh_default_org/internal/69f554e6-98ad-4e60-ad3b-ff8cdc00a887/7d9e16a2-114a-4bf9-86fc-cc506236659b/composite.mp4' from origin 'https://<admin-node>' has been blocked by CORS policy: No 'Access-Control-Allow-Origin' header is present on the requested resource.
composite.mp4:1          Failed to load resource: net::ERR_FAILED
composite.mp4:1          Failed to load resource: the server responded with a status of 403 (Forbidden)
composite.mp4:1          Failed to load resource: the server responded with a status of 403 (Forbidden)
/admin-ng/index.html;jsessionid=node0ubq5wme8kduk148dsxtqqovuu2517.node0:1          Failed to load resource: net::ERR_TOO_MANY_REDIRECTS
```

**To Reproduce**
1. open Admin UI
2. use "open video editor" button on any video
3. waveform gets the "wheel of death"
4. open browser console for errors

**Environment Information**
- Cluster: 1 admin vm, 1 worker vm, 1 presentation vm
- OS: CentOS 7 (Red Hat 4.8.5-39)
- MariaDB Ver 10.7.3
- Shibboleth Auth
- Apache httpd
- Browser: Chrome, Firefox, Edge
- Opencast 11.4

**Our Fix:**
Add a line to the apache ssl file `/etc/httpd/conf.d/<presentation-node-url>.ssl.conf`:
```
Header add Access-Controll-Allow-Origin <admin – server – node – adress>
```

**Comment by Rillke** [comment]
> CORS missing allow origin

Is the .4 version cursed? I can see this issue in 12.4 now, too although I am sure, I didn't experience that in 12.3 where we ran the editor without any -UI-configuration. Wondering if something changed in the editor or if the added configuration did that.

![Screenshot from 2022-10-21 17-16-29](https://user-images.githubusercontent.com/2311611/197231364-f050590b-1f92-4fe7-b8cd-0f4d451b987b.png)

The waveform is not an issue because it's fetched as an img src.

**Comment by Rillke** [comment]
BTW, we are redirecting requests coming from the editors to the internal publication channel to the presentation node back to the admin node because a user might not be authenticated at the presentation server. I wonder if we misconfigured something.

```
# preview videos should be served by admin
# where user is authenticated when editing videos
location /static/mh_default_org/internal {
    return 302 $admin_url$request_uri;
}
```

**Comment by mtneug** [comment]
What is still missing with #4411 now in?

IMO this is more a proxy config problem?

**Comment by Rillke** [comment]
> IMO this is more a proxy config problem?

No, not really. This is still an issue. Although `"local": true`, the UI still requests to our presentation/engage while sitting on our admin.

![image](https://user-images.githubusercontent.com/2311611/205093034-2c5506fc-53b5-49f4-a41c-06e3d8eb993c.png)

The issue here is authentication. Someone working on the admin might not be authenticated at the presentation. This is fine for everything that is not XHR-loaded because a simple redirect back to the admin just works. With XHR it gets more complicated though: The java session cookie is not sent along the request to another origin if not explicitly set while making the request (the editor probably does not set withCredentials) and you need to answer with a proper `access-control-allow-credentials` (we do this) and even though our redirect back to the admin works, it arrives at our admin without credentials (because the editor didn't do the request withCredentials) and [with an origin of `null`](https://stackoverflow.com/questions/30193851/ajax-call-following-302-redirect-sets-origin-to-null) in Firefox.

![image](https://user-images.githubusercontent.com/2311611/205095511-277d4077-d235-41b9-864b-04b915f74a6c.png)

So I guess we have to wait until the editor honors the `"local": true` response from  #4411

**Comment by mtneug** [comment]
CORS headers can be added by the proxy, thus my comment. If media delivery is cookie protected, then yes, but this is another issue.

**Comment by Rillke** [comment]
We have enabled cookie based file delivery protection (aka. securing static files). Would you like me to [file a separate issue against the editor](https://github.com/opencast/opencast-editor/)?

**Comment by Rillke** [comment]
#1010 was merged, which resolves the issue for us.

**Comment by lkiesow** [comment]
Sorry, I missed this issue when filing the pull request against the editor.
I'll try to get a new version of the editor in soonish.

In the meantime, if you run into this issue, you can fix this in Opencast with a little bit of Nginx magic _on the admin node_:

```py
    location ~ /editor/.*/edit.json {

      proxy_set_header Host $host;
      proxy_set_header X-Real-IP $remote_addr;
      proxy_set_header X-Forwarded-For $proxy_add_x_forwarded_for;
      proxy_set_header X-Forwarded-Proto $scheme;
      proxy_set_header X-Forwarded-SSL "on";

      proxy_pass http://127.0.0.1:8080;

      # additional CORS and proxy settings here

      sub_filter_once off;
      sub_filter_types application/json;
      sub_filter 'presentation.your-opencast.org' 'admin.your-opencast.org';
    }
```

This will filter the response from Opencast to the video editor and replace the presentation links with admin links
